Increase test coverage of rcl_action

brawner opened the pull request:
This boosts the coverage of rcl_action tests. 

* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=10919)](http://ci.ros2.org/job/ci_linux/10919/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=6269)](http://ci.ros2.org/job/ci_linux-aarch64/6269/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=8891)](http://ci.ros2.org/job/ci_osx/8891/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=10820)](http://ci.ros2.org/job/ci_windows/10820/)

brawner commented:
Force pushed white-space fix for uncrustify error.
[![Build Status](http://build.ros2.org/buildStatus/icon?job=Fpr__rcl__ubuntu_focal_amd64&build=93)](http://build.ros2.org/job/Fpr__rcl__ubuntu_focal_amd64/93/)

jacobperron commented on rcl_action/test/rcl_action/test_action_server.cpp:
```diff
@@ -135,6 +168,14 @@ TEST(TestActionServerInitFini, test_action_server_init_fini)
   ret = rcl_action_server_fini(&action_server, &node);
   EXPECT_EQ(ret, RCL_RET_OK) << rcl_get_error_string().str;
 
+  // Setup again for other finalize checks
```
It looks like this case is already covered by the by the initialization at L143 and the fini at L168, no?

jacobperron commented on rcl_action/test/rcl_action/test_action_server.cpp:
```diff
@@ -28,6 +28,39 @@
+}
+
+/// This is fully defined in action_server.c, reproducing here for testing.
+typedef struct rcl_action_server_impl_t
```
This seems brittle since we'll have to remember to update this if the copy in action_server.c changes.
Can we move the definition in action_server.c to a private header that can be included here as well?

brawner replied:
Moved these structs to an impl directory in include

jacobperron commented on rcl_action/test/rcl_action/test_action_server.cpp:
```diff
@@ -308,6 +389,74 @@ TEST_F(TestActionServer, test_action_accept_new_goal)
+  init_test_uuid1(goal_info_in.goal_id.uuid);
+
+  // Check doesn't exist
+  EXPECT_FALSE(rcl_action_server_goal_exists(&this->action_server, &different_goal));
```
If we're checking `rcl_error_is_set()` in the failure cases, I suppose it also makes sense to check in the nominal cases. E.g.

```c++
EXPECT_FALSE(rcl_error_is_set());
```

brawner replied:
Done.

jacobperron commented on rcl_action/test/rcl_action/test_wait.cpp:
```diff
@@ -0,0 +1,847 @@
+#include "osrf_testing_tools_cpp/scope_exit.hpp"
+#include "test_msgs/action/fibonacci.h"
+
+/// This is fully defined in action_server.c, reproducing here for testing.
```
same comment here about removing duplication.

brawner replied:
Done.

jacobperron commented on rcl_action/test/rcl_action/test_wait.cpp:
```diff
@@ -0,0 +1,847 @@
+  size_t wait_set_expire_timer_index;
+} rcl_action_server_impl_t;
+
+/// This is fully defined in action_client.c, reproducing here for testing.
```
same comment here about removing duplication.

brawner replied:
Done.

jacobperron commented on rcl_action/src/rcl_action/action_server.c:
```diff
@@ -693,6 +693,9 @@ rcl_ret_t
 rcl_action_notify_goal_done(
   const rcl_action_server_t * action_server)
 {
+  if (!rcl_action_server_is_valid(action_server)) {
```
I assume this is to fix one of the new tests added? It's probably worth mentioning in the commit message that this PR contains a bug fix as well.

brawner replied:
Separated this and the changes to goal_state_machine.c into their own commit.

jacobperron commented:
I'm still looking over `test_wait.cpp`.

jacobperron commented on rcl_action/test/rcl_action/test_wait.cpp:
```diff
@@ -0,0 +1,847 @@
+  const rcl_action_client_options_t action_client_options =
+    rcl_action_client_get_default_options();
+
+  rcl_action_client_t action_client = rcl_action_get_zero_initialized_client();
```
Can we move the init/fini of the action client to the test fixture (like we did for the server)?

brawner commented:
Addressed PR feedback and rerunning tests:

* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=10924)](http://ci.ros2.org/job/ci_linux/10924/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=6274)](http://ci.ros2.org/job/ci_linux-aarch64/6274/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=8896)](http://ci.ros2.org/job/ci_osx/8896/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=10826)](http://ci.ros2.org/job/ci_windows/10826/)

jacobperron commented:
Look like `test_server_wait_set_get_entities_ready` isn't working on macOS and Windows.

jacobperron commented on rcl_action/include/rcl_action/impl/action_client.h:
```diff
@@ -0,0 +1,39 @@
+#include "rcl_action/types.h"
+#include "rcl/rcl.h"
+
+typedef struct rcl_action_client_impl_t
```
Is it possible to move the implementation headers to the `src` directory? This way we avoid installing them and having users misuse them.

If not, we can leave it as-is.

brawner replied:
Moved

brawner commented:
I think I took care of the memory error. Checking again:

* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=10931)](http://ci.ros2.org/job/ci_linux/10931/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=6278)](http://ci.ros2.org/job/ci_linux-aarch64/6278/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=8900)](http://ci.ros2.org/job/ci_osx/8900/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=10832)](http://ci.ros2.org/job/ci_windows/10832/)

brawner commented:
About 20 lines short (out of 1030) to reach 95% coverage. Guess that will have to do for now.

The automated build.ros2.org check failed when I pushed the most recent commit because the nodes in rcl-action interfered with rcl's tests. Not sure why since they are run sequentially. I force pushed again and it passed. Hopefully this doesn't show up regularly in nightlies.

ivanpauno commented:
This introduced 11 new build warnings in nightlies https://ci.ros2.org/view/nightly/job/nightly_linux_release/1557/gcc/new/

ivanpauno commented on rcl_action/src/rcl_action/goal_state_machine.c:
```diff
@@ -26,51 +26,40 @@ typedef rcl_action_goal_state_t
-    return GOAL_STATE_UNKNOWN;
-  }
+  assert(GOAL_STATE_ACCEPTED == state || GOAL_STATE_EXECUTING == state);
+  assert(GOAL_EVENT_CANCEL_GOAL == event);
```
assert is compiled out in release mode, that's why the new warnings.

why a PR titled "Increase test coverage ..." is changing behavior?
If a behavior change was needed, that should've been part of a different PR (or at least, commits shouldn't have been squashed)

brawner commented:
I’ll submit a patch soon. This doesn’t actually change behavior, the if statements were checking conditions which were impossible based on logic elsewhere.

> On May 29, 2020, at 7:07 AM, Ivan Santiago Paunovic <notifications@github.com> wrote:
> 
> ﻿
> @ivanpauno commented on this pull request.
> 
> In rcl_action/src/rcl_action/goal_state_machine.c:
> 
> >    return GOAL_STATE_EXECUTING;
>  }
>  
>  rcl_action_goal_state_t
>  _cancel_goal_event_handler(rcl_action_goal_state_t state, rcl_action_goal_event_t event)
>  {
> -  if ((GOAL_STATE_ACCEPTED != state && GOAL_STATE_EXECUTING != state) ||
> -    GOAL_EVENT_CANCEL_GOAL != event)
> -  {
> -    return GOAL_STATE_UNKNOWN;
> -  }
> +  assert(GOAL_STATE_ACCEPTED == state || GOAL_STATE_EXECUTING == state);
> +  assert(GOAL_EVENT_CANCEL_GOAL == event);
> assert is compiled out in release mode, that's why the new warnings.
> 
> why a PR titled "Increase test coverage ..." is changing behavior?
> If a behavior change was needed, that should've been part of a different PR (or at least, commits shouldn't have been squashed)
> 
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub, or unsubscribe.

ivanpauno commented:
> This doesn’t actually change behavior, the if statements were checking conditions which were impossible based on logic elsewhere.

Sounds good to me then!

brawner commented:
Addressed in https://github.com/ros2/rcl/pull/666
